src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java

```diff
@@ -161,14 +162,95 @@ public boolean isUpdateSuccessful() {
       .thenCompose(jsonObj -> updateOrderLine(compPOL, jsonObj))
       .thenCompose(v -> inventoryHelper.handleHoldingRecord(compPOL))
       .thenCompose(holdingId -> inventoryHelper.handleItemRecords(compPOL, holdingId))
-      .thenAccept(itemIds -> {
-        // Temporal check. The idea is to create piece records for successfully created items and then throw exception
+      .thenCompose(itemIds -> {
         if (itemIds.size() != expectedItemsQuantity) {
           throw new IllegalStateException("Expected items quantity does not correspond to created items");
         }
-      });
+        return completedFuture(itemIds);
+      })
+      .thenCompose(itemIds -> createPieces(compPOL, expectedItemsQuantity, itemIds));
+  }
+
+  /**
+   * Creates Piece records corresponding to each item record associated with PO Line
+   *
+   * @param compPOL Composite PO line to update Inventory for
+   * @param expectedItemsQuantity Inventory items quantity
+   * @param itemIds List of item id
+   * @return CompletableFuture with newly created Pieces associated with PO line.
+   */
+  private CompletableFuture<Void> createPieces(PoLine compPOL, int expectedItemsQuantity, List<String> itemIds) {
+		CompletableFuture<Void> future = new VertxCompletableFuture<>(ctx);
+		if (itemIds.isEmpty()) {
+			future.complete(null);
+			return future;
+		}
+		List<CompletableFuture<Void>> futuresList = new ArrayList<>();
+		String poLineId = compPOL.getId();		
+		String endpoint = String.format(PIECES_ENDPOINT, poLineId);
+		
+		handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+		    .thenAccept(body -> {
+		    	int i=0;
+		      // No Pieces exists
+			    if (body.getInteger("total_records") == 0) { 
```

Comment:
At the moment there is no filtering logic and basically we cannot rely on the order of the item identifiers (because it is fully depends on Inventory storage implementation). So the logic might always filter items based on existing piece records and create piece records for remaining item id's. Something like:
```java
          PieceCollection pieces = body.mapTo(PieceCollection.class);
          List<String> existingItemIds = pieces.getPieces()
                                               .stream()
                                               .map(Piece::getItemId)
                                               .filter(StringUtils::isNotEmpty)
                                               .collect(Collectors.toList());
          itemIds.stream()
                 .filter(id -> !existingItemIds.contains(id))
                 .forEach(itemId -> futuresList.add(createPiece(poLineId, itemId)));
```

Reply:
Done. Nice recommendation! Thanks @piotr-kalashuk